soweego/wikidata/api_requests.py

```diff
@@ -36,9 +34,272 @@
+    needs_publication_date = entity in constants.REQUIRE_PUBLICATION_DATE
+
+    # Initialize 7 counters to 0
+    # Indices legend:
+    # 0 = claims
+    # 1 = labels
+    # 2 = aliases
+    # 3 = descriptions
+    # 4 = sitelinks
+    # 5 = third-party URLs
+    # 6 = third-party IDs
+    counters = [0] * 7
+
+    # Create a partial function where all parameters
+    # but the data bucket are passed to `_process_bucket`,
+    # so that we only pass the data bucket
+    # when we call `pool_function`.
+    # In this way, it becomes trivial to use
+    # `multiprocessing.Pool` map functions, like `imap_unordered`
+    pool_function = partial(
+        _process_bucket,
+        request_params=request_params,
+        url_pids=url_pids,
+        ext_id_pids_to_urls=ext_id_pids_to_urls,
+        qids_and_tids=qids_and_tids,
+        needs=(needs_occupation, needs_genre, needs_publication_date),
+        counters=counters,
+    )
+
+    # Create a pool of threads and map the list of buckets via `pool_function`
+    with Pool() as pool:
+        # `processed_bucket` will be a list of dicts, where each dict
+        # is a processed entity from the bucket
+        for processed_bucket in pool.imap_unordered(
+            pool_function, tqdm(qid_buckets, total=len(qid_buckets))
+        ):
+            # Join results into a string so that we can write them to
+            # the dump file
+            to_write = ''.join(
+                json.dumps(result, ensure_ascii=False) + '\n'
+                for result in processed_bucket
+            )
+
+            fileout.write(to_write)
+            fileout.flush()
+
+    LOGGER.info(
+        'QIDs: got %d with no expected claims, %d with no labels, '
+        '%d with no aliases, %d with no descriptions, %d with no sitelinks, '
+        '%d with no third-party links, %d with no external ID links',
+        *counters
+    )
+
+
+@lru_cache()
+def build_session() -> requests.Session:
+    """Build the HTTP session for interaction with the Wikidata API.
+
+    Log in if credentials are found,
+    otherwise go ahead with an unauthenticated session.
+    If a previously cached session has expired, build a new one.
+
+    :rtype: :py:class:`requests.Session`
+    :return: the HTTP session to interact with the Wikidata API
+    """
+    session_dump_path = os.path.join(
+        constants.SHARED_FOLDER, constants.WIKIDATA_API_SESSION
+    )
+
+    try:
+        return _load_cached_session(session_dump_path)
+
+    except (FileNotFoundError, AssertionError):
+        LOGGER.debug('Logging into the Wikidata API ...')
+        try:
+            # Try to login by loading credentials from file
+            success, err_msg, session = _login(*_get_credentials_from_file())
+
+            # Login failed: wrong user and/or password.
+            # Stop execution
+            if not success:
+                raise AssertionError(err_msg)
+
+        except (FileNotFoundError, KeyError) as error:
```

Comment:
A cleaner way to do this would be to catch different exception types:

``` python
except FileNotFound:
    LOGGER.info(...)

except KeyError:
    LOGGER.info(...)
```

Reply:
I've already tried first what you suggest, but with 2 `except` clauses, we have to repeat the `global` statement, and this makes the Python interpreter cry.

Reply:
good point. well in that case I'll mark this as resolved 😄